File: crates/actant-contracts/src/policy.rs
//! Policy types: rules, verdicts, approvals.

use schemars::JsonSchema;
use serde::{Deserialize, Serialize};

use crate::Risk;

/// A Guard verdict on a proposed tool call.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, JsonSchema)]
#[serde(tag = "decision", rename_all = "snake_case")]
pub enum PolicyVerdict {
    /// Allow the call as proposed.
    Allow {
        /// Human-readable reason for Studio display.
        reason: String,
        /// Snapshot hash of the policy that produced this verdict.
        policy_snapshot: String,
    },
    /// Allow a rewritten variant of the call.
    Constrain {
        /// Reason.
        reason: String,
        /// Snapshot hash.
        policy_snapshot: String,
        /// Rewritten arguments to use in place of the original.
        constrained_input: serde_json::Value,
        /// Short hint shown to the approver if approval is requested.
        hint: String,
    },
    /// Require explicit approval before the call executes.
    RequireApproval {
        /// Reason.
        reason: String,
        /// Snapshot hash.
        policy_snapshot: String,
        /// Optional constrain hint to offer the approver.
        #[serde(default, skip_serializing_if = "Option::is_none")]
        hint: Option<String>,
        /// Optional constrained args the approver can choose to accept.
        #[serde(default, skip_serializing_if = "Option::is_none")]
        constrained_input: Option<serde_json::Value>,
    },
    /// Block the call. The run continues but the tool call does not execute.
    Block {
        /// Reason.
        reason: String,
        /// Snapshot hash.
        policy_snapshot: String,
    },
    /// Halt the entire agent run.
    Halt {
        /// Reason.
        reason: String,
        /// Snapshot hash.
        policy_snapshot: String,
    },
}

impl PolicyVerdict {
    /// Returns a stable lowercase kind (for ledger payloads and matching).
    pub fn kind(&self) -> &'static str {
        match self {
            Self::Allow { .. } => "allow",
            Self::Constrain { .. } => "constrain",
            Self::RequireApproval { .. } => "require_approval",
            Self::Block { .. } => "block",
            Self::Halt { .. } => "halt",
        }
    }
}

/// A simple regex-based deny rule applied to tool arguments.
#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema)]
pub struct ArgDenyRule {
    /// Tool name this rule applies to (e.g. "shell.run"). "*" for any tool.
    pub tool: String,
    /// Regular expression evaluated against the JSON-stringified args.
    pub pattern: String,
    /// Human reason; surfaced in the verdict.
    pub reason: String,
}

/// Per-tool risk classification entry.
#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema)]
pub struct ToolRiskEntry {
    /// Tool name.
    pub tool: String,
    /// Risk level assigned by the policy.
    pub risk: Risk,
    /// If true, every call requires approval regardless of args.
    #[serde(default)]
    pub require_approval: bool,
}

/// The v0.1 policy document. Small, opinionated, easy to read.
#[derive(Debug, Clone, Default, Serialize, Deserialize, JsonSchema)]
pub struct Policy {
    /// Per-tool risk class. Default risk if missing: `low`.
    #[serde(default)]
    pub tools: Vec<ToolRiskEntry>,
    /// Regex deny-list applied to args before risk evaluation.
    #[serde(default)]
    pub deny: Vec<ArgDenyRule>,
    /// Highest sensitivity allowed without approval.
    #[serde(default)]
    pub sensitivity_ceiling: Option<crate::Sensitivity>,
    /// Free-text label for Studio display.
    #[serde(default)]
    pub label: String,
}

/// A pending or recorded approval request.
#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema)]
pub struct ApprovalRequest {
    /// Tool call id this approval is for.
    pub tool_call_id: String,
    /// Tool name.
    pub tool: String,
    /// Original arguments (pre-constrain).
    pub args: serde_json::Value,
    /// Constrain hint, if Guard suggested one.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub hint: Option<String>,
    /// Constrained args the approver can accept.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub constrained_input: Option<serde_json::Value>,
    /// Reason from Guard.
    pub reason: String,
}

/// Approval outcome.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, JsonSchema)]
#[serde(tag = "decision", rename_all = "snake_case")]
pub enum ApprovalDecision {
    /// Approved as-proposed.
    Approve {
        /// Approver identifier.
        approver: String,
        /// Approval scope (e.g. "once", "session").
        scope: String,
    },
    /// Approved the constrained variant.
    ApproveConstrained {
        /// Approver identifier.
        approver: String,
        /// Approval scope.
        scope: String,
        /// Final arguments accepted.
        accepted_input: serde_json::Value,
    },
    /// Denied.
    Deny {
        /// Approver identifier.
        approver: String,
        /// Free-text reason.
        reason: String,
    },
}

File: sdks/swift/Sources/ActantDB/Types/Policy.swift
import Foundation

// MARK: - Policy document

public struct ToolRiskEntry: Codable, Sendable {
    public let tool: String
    public let risk: Risk
    public let requireApproval: Bool

    public init(tool: String, risk: Risk, requireApproval: Bool = false) {
        self.tool = tool; self.risk = risk; self.requireApproval = requireApproval
    }

    enum CodingKeys: String, CodingKey {
        case tool, risk
        case requireApproval = "require_approval"
    }

    public init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        self.tool = try c.decode(String.self, forKey: .tool)
        self.risk = try c.decode(Risk.self, forKey: .risk)
        self.requireApproval = try c.decodeIfPresent(Bool.self, forKey: .requireApproval) ?? false
    }
}

public struct ArgDenyRule: Codable, Sendable {
    public let tool: String
    public let pattern: String
    public let reason: String

    public init(tool: String, pattern: String, reason: String) {
        self.tool = tool; self.pattern = pattern; self.reason = reason
    }
}

public struct Policy: Codable, Sendable {
    public let tools: [ToolRiskEntry]
    public let deny: [ArgDenyRule]
    public let sensitivityCeiling: Sensitivity?
    public let label: String

    public init(
        tools: [ToolRiskEntry] = [],
        deny: [ArgDenyRule] = [],
        sensitivityCeiling: Sensitivity? = nil,
        label: String = ""
    ) {
        self.tools = tools; self.deny = deny
        self.sensitivityCeiling = sensitivityCeiling; self.label = label
    }

    enum CodingKeys: String, CodingKey {
        case tools, deny
        case sensitivityCeiling = "sensitivity_ceiling"
        case label
    }

    public init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        self.tools = try c.decodeIfPresent([ToolRiskEntry].self, forKey: .tools) ?? []
        self.deny  = try c.decodeIfPresent([ArgDenyRule].self,  forKey: .deny)  ?? []
        self.sensitivityCeiling = try c.decodeIfPresent(Sensitivity.self, forKey: .sensitivityCeiling)
        self.label = try c.decodeIfPresent(String.self, forKey: .label) ?? ""
    }
}

// MARK: - PolicyVerdict (internally-tagged union, custom Codable)
//
// Rust shape: `#[serde(tag = "decision", rename_all = "snake_case")]`
// JSON: { "decision": "allow", "reason": "...", "policy_snapshot": "..." }

public enum PolicyVerdict: Codable, Sendable, Equatable {
    case allow(reason: String, policySnapshot: String)
    case constrain(reason: String, policySnapshot: String,
                   constrainedInput: JSONValue, hint: String)
    case requireApproval(reason: String, policySnapshot: String,
                         hint: String?, constrainedInput: JSONValue?)
    case block(reason: String, policySnapshot: String)
    case halt(reason: String, policySnapshot: String)

    /// Stable wire kind (matches Rust `PolicyVerdict::kind()`).
    public var kind: String {
        switch self {
        case .allow:           return "allow"
        case .constrain:       return "constrain"
        case .requireApproval: return "require_approval"
        case .block:           return "block"
        case .halt:            return "halt"
        }
    }

    private enum K: String, CodingKey {
        case decision, reason
        case policySnapshot   = "policy_snapshot"
        case constrainedInput = "constrained_input"
        case hint
    }

    public init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: K.self)
        let decision = try c.decode(String.self, forKey: .decision)
        let reason   = try c.decode(String.self, forKey: .reason)
        let snap     = try c.decode(String.self, forKey: .policySnapshot)
        switch decision {
        case "allow":
            self = .allow(reason: reason, policySnapshot: snap)
        case "constrain":
            let input = try c.decode(JSONValue.self, forKey: .constrainedInput)
            let hint  = try c.decode(String.self,    forKey: .hint)
            self = .constrain(reason: reason, policySnapshot: snap,
                              constrainedInput: input, hint: hint)
        case "require_approval":
            let hint  = try c.decodeIfPresent(String.self,    forKey: .hint)
            let input = try c.decodeIfPresent(JSONValue.self, forKey: .constrainedInput)
            self = .requireApproval(reason: reason, policySnapshot: snap,
                                    hint: hint, constrainedInput: input)
        case "block":
            self = .block(reason: reason, policySnapshot: snap)
        case "halt":
            self = .halt(reason: reason, policySnapshot: snap)
        default:
            throw DecodingError.dataCorruptedError(
                forKey: .decision, in: c,
                debugDescription: "unknown PolicyVerdict decision: \(decision)"
            )
        }
    }

    public func encode(to encoder: Encoder) throws {
        var c = encoder.container(keyedBy: K.self)
        try c.encode(kind, forKey: .decision)
        switch self {
        case let .allow(reason, snap), let .block(reason, snap), let .halt(reason, snap):
            try c.encode(reason, forKey: .reason)
            try c.encode(snap,   forKey: .policySnapshot)
        case let .constrain(reason, snap, input, hint):
            try c.encode(reason, forKey: .reason)
            try c.encode(snap,   forKey: .policySnapshot)
            try c.encode(input,  forKey: .constrainedInput)
            try c.encode(hint,   forKey: .hint)
        case let .requireApproval(reason, snap, hint, input):
            try c.encode(reason, forKey: .reason)
            try c.encode(snap,   forKey: .policySnapshot)
            try c.encodeIfPresent(hint,  forKey: .hint)
            try c.encodeIfPresent(input, forKey: .constrainedInput)
        }
    }
}

// MARK: - Approvals

public struct ApprovalRequest: Codable, Sendable {
    public let toolCallID: String
    public let tool: String
    public let args: JSONValue
    public let hint: String?
    public let constrainedInput: JSONValue?
    public let reason: String

    enum CodingKeys: String, CodingKey {
        case toolCallID       = "tool_call_id"
        case tool, args, hint
        case constrainedInput = "constrained_input"
        case reason
    }
}

/// Approval outcome. Wire: `{"decision": "approve" | "approve_constrained" | "deny", ...}`.
public enum ApprovalDecisionV: Codable, Sendable, Equatable {
    case approve(approver: String, scope: String)
    case approveConstrained(approver: String, scope: String, acceptedInput: JSONValue)
    case deny(approver: String, reason: String)

    public var kind: String {
        switch self {
        case .approve:            return "approve"
        case .approveConstrained: return "approve_constrained"
        case .deny:               return "deny"
        }
    }

    private enum K: String, CodingKey {
        case decision, approver, scope, reason
        case acceptedInput = "accepted_input"
    }

    public init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: K.self)
        let decision = try c.decode(String.self, forKey: .decision)
        let approver = try c.decode(String.self, forKey: .approver)
        switch decision {
        case "approve":
            let scope = try c.decode(String.self, forKey: .scope)
            self = .approve(approver: approver, scope: scope)
        case "approve_constrained":
            let scope = try c.decode(String.self, forKey: .scope)
            let input = try c.decode(JSONValue.self, forKey: .acceptedInput)
            self = .approveConstrained(approver: approver, scope: scope, acceptedInput: input)
        case "deny":
            let reason = try c.decode(String.self, forKey: .reason)
            self = .deny(approver: approver, reason: reason)
        default:
            throw DecodingError.dataCorruptedError(
                forKey: .decision, in: c,
                debugDescription: "unknown ApprovalDecision: \(decision)"
            )
        }
    }

    public func encode(to encoder: Encoder) throws {
        var c = encoder.container(keyedBy: K.self)
        try c.encode(kind, forKey: .decision)
        switch self {
        case let .approve(approver, scope):
            try c.encode(approver, forKey: .approver)
            try c.encode(scope,    forKey: .scope)
        case let .approveConstrained(approver, scope, input):
            try c.encode(approver, forKey: .approver)
            try c.encode(scope,    forKey: .scope)
            try c.encode(input,    forKey: .acceptedInput)
        case let .deny(approver, reason):
            try c.encode(approver, forKey: .approver)
            try c.encode(reason,   forKey: .reason)
        }
    }
}

/// Pending-approval row returned by `GET /v1/approvals?workspace_id=...`.
/// Distinct from `ApprovalRequest` (Guard's contract payload).
public struct PendingApproval: Codable, Sendable, Identifiable {
    public let id: String
    public let toolCallID: String
    public let requestedBy: String
    public let riskLevel: String
    public let summary: String
    public let status: String

    enum CodingKeys: String, CodingKey {
        case id
        case toolCallID  = "tool_call_id"
        case requestedBy = "requested_by"
        case riskLevel   = "risk_level"
        case summary, status
    }
}
